Userland/Services/FileSystemAccessServer/ClientConnection.cpp

```diff
@@ -0,0 +1,166 @@
+#include <LibCore/StandardPaths.h>
+#include <LibGUI/Application.h>
+#include <LibGUI/FilePicker.h>
+#include <LibGUI/MessageBox.h>
+
+namespace FileSystemAccessServer {
+
+static HashMap<int, RefPtr<ClientConnection>> s_connections;
+
+ClientConnection::ClientConnection(NonnullRefPtr<Core::LocalSocket> socket, int client_id)
+    : IPC::ClientConnection<FileSystemAccessClientEndpoint, FileSystemAccessServerEndpoint>(*this, move(socket), client_id)
+{
+    s_connections.set(client_id, *this);
+}
+
+ClientConnection::~ClientConnection()
+{
+}
+
+void ClientConnection::die()
+{
+    s_connections.remove(client_id());
+    GUI::Application::the()->quit();
+    exit(0);
+}
+
+Messages::FileSystemAccessServer::RequestFileResponse ClientConnection::request_file(String const& sender, String const& path, int requested_access)
+{
+    // TODO: This hangs the calling application
+
+    // Must receive an absolute path
+    VERIFY(path.starts_with("/"sv));
+
+    bool approved = false;
+    if (m_approved_files.get(path).has_value()) {
+        approved = (requested_access & m_approved_files.get(path).value()) == requested_access;
+    }
+
+    if (!approved) {
+        StringBuilder builder;
+        if (requested_access & O_RDONLY)
+            builder.append('r');
+
+        if (requested_access & O_WRONLY)
+            builder.append('w');
+
+        if (requested_access & O_CREAT)
+            builder.append('c');
+
+        auto access_string = builder.to_string();
+
+        auto result = GUI::MessageBox::show(nullptr, String::formatted("Give {} \"{}\" access to \"{}\"?", sender, access_string, path), "File Permissions Requested", GUI::MessageBox::Type::Warning, GUI::MessageBox::InputType::YesNo);
+
+        approved = result == GUI::MessageBox::ExecYes;
+
+        if (approved) {
+            // Intentionally overwrite existing access
+            m_approved_files.set(path, requested_access);
+        }
+    }
+
+    if (approved) {
+        auto fd = open(path.characters(), requested_access);
+
+        if (fd < 0) {
+            dbgln(String::formatted("FileSystemAccessServer: Couldn't open {}, error {}", path, OSError(errno).string()));
+
+            return { errno, Optional<IPC::File> {} };
+        }
+
+        return { 0, IPC::File(fd, IPC::File::CloseAfterSending) };
+    }
+
+    return { -1, Optional<IPC::File> {} };
+}
+
+Messages::FileSystemAccessServer::PromptOpenFileResponse ClientConnection::prompt_open_file(String const& path_to_view, int requested_access)
+{
+    auto main_window = GUI::Window::construct();
+    Optional<String> user_picked_file = GUI::FilePicker::get_open_filepath(main_window, "Select file", path_to_view);
+
+    if (user_picked_file.has_value()) {
+        auto fd = open(user_picked_file->characters(), requested_access, 0644);
+
+        if (fd < 0) {
+            dbgln(String::formatted("FileSystemAccessServer: Couldn't open {}, error {}", user_picked_file.value(), OSError(errno).string()));
```

Comment:
Don't use `String::formatted` inside `dbgln`. Firstly, it's unnecessary since `dbgln` already supports formatting. Secondly, if the formatted String still contains curly braces, `dbgln` will interpret them, you you would need to put `dbgln("{}", String::formatted...`. But what you actually need to do it this:
```suggestion
            dbgln("FileSystemAccessServer: Couldn't open {}, error {}", user_picked_file.value(), OSError(errno).string());
```